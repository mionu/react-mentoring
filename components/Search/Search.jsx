import React, { useCallback, useState } from 'react';
import { useRouter } from 'next/router';
import { TextField, Button } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';

const useStyles = makeStyles({
  search: {
    display: 'flex',
    justifyContent: 'center',
  },
  searchInput: {
    flexGrow: 2,
    marginRight: 10,
  },
});

const Search = () => {
  const classes = useStyles();
  const router = useRouter();
  const { query } = router.query;
  const defaultQuery = query || '';
  const [term, setSearchTerm] = useState(defaultQuery);

  const handleInput = useCallback(({ target }) => {
    setSearchTerm(target.value);
  }, []);

  const runSearch = () => {
    router.push(!term ? '/' : `/search/${term}`);
  };

  const handleKeyPress = useCallback(({ key }) => {
    if (key === 'Enter') {
      runSearch();
    }
  }, []);

  return (
    <>
      <h1>FIND YOUR MOVIE</h1>
      <div className={classes.search}>
        <TextField
          className={classes.searchInput}
          size="small"
          variant="outlined"
          label="What do you want to watch?"
          value={term}
          onChange={handleInput}
          onKeyPress={handleKeyPress}
        />
        <Button variant="contained" onClick={runSearch}>Search</Button>
      </div>
    </>
  );
};

export default Search;
