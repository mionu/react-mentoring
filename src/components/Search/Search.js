import React, { useState } from 'react';
import { TextField, Button } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import { connect } from 'react-redux';
import { setOptions } from '../../redux/actions/action-creators';

const useStyles = makeStyles({
    search: {
        display: 'flex',
        justifyContent: 'center',
    },
    searchInput: {
        flexGrow: 2,
        marginRight: 10,
    },
});

const Search = (props) => {
    const classes = useStyles();
    const [searchTerm, setSearchTerm] = useState(props.searchTerm || '');

    const handleInput = ({ target }) => {
        setSearchTerm(target.value);
    };

    const handleKeyPress = ({ key }) => {
        if (key === 'Enter') {
            runSearch();
        }
    }
    
    const runSearch = () => {
        props.search(searchTerm);
    };

    return <>
        <h1>FIND YOUR MOVIE</h1>
        <div className={classes.search}>
            <TextField
                className={classes.searchInput}
                size='small'
                variant='outlined'
                label='What do you want to watch?'
                value={searchTerm}
                onChange={handleInput}
                onKeyPress={handleKeyPress}
            />
            <Button variant='contained' onClick={runSearch}>Search</Button>
        </div>
    </>;
}

const mapStateToProps = ({ moviesReducer }) => ({ searchTerm: moviesReducer.options.search });

const mapDispatchToProps = (dispatch) => ({
    search: (search) => dispatch(setOptions({ search })),
});

export default connect(mapStateToProps, mapDispatchToProps)(Search);
