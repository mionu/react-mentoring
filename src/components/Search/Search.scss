.search {
    display: flex;
    justify-content: center;

    .search-input {
        flex-grow: 2;
        margin-right: 10px;
    }
}