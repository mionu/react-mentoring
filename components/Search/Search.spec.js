import React from 'react';
import { RouterContext } from 'next/dist/next-server/lib/router-context';
import renderer from 'react-test-renderer';
import { TextField } from '@material-ui/core';
import Search from './Search';

/* eslint-disable no-undef, react/jsx-filename-extension */
describe('Search', () => {
  beforeAll(() => jest.spyOn(React, 'useEffect').mockImplementation(React.useLayoutEffect));
  afterAll(() => React.useEffect.mockRestore());

  it('should render', () => {
    const searchTerm = 'search';

    const component = renderer.create(
      <RouterContext.Provider value={{ query: { query: searchTerm } }}>
        <Search />
      </RouterContext.Provider>,
    );

    expect(component.toJSON).toMatchSnapshot();

    const search = component.root.findByType(TextField);
    expect(search).toBeDefined();
    expect(search.props.value).toBe(searchTerm);
  });
});
/* eslint-enable no-undef, react/jsx-filename-extension */
