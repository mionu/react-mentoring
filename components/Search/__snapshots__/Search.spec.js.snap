// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search should render 1`] = `[Function]`;
